File: frontend/src/lib/odoo.ts
export interface OdooHealth {
  connected: boolean;
  read_only: boolean;
  version?: string | null;
  database?: string | null;
  message: string;
  active_company_id?: number | null;
  active_company_name?: string | null;
}

export interface OdooCompany {
  id: number;
  name: string;
  currency_id?: number | null;
  currency_name?: string | null;
  partner_id?: number | null;
  is_active: boolean;
  selected_by_current_user: boolean;
}

export interface OdooCompaniesResponse {
  items: OdooCompany[];
  connected: boolean;
  message?: string | null;
}

export interface OdooUserMapping {
  id?: string | null;
  tenant_id: string;
  jaios_user_id: string;
  odoo_user_id: number;
  odoo_login: string;
  odoo_partner_id?: number | null;
  allowed_company_ids: number[];
  default_company_id?: number | null;
  is_active: boolean;
  last_verified_at?: string | null;
}

export interface OdooMe {
  jaios_user_id: string;
  jaios_email: string;
  jaios_name: string;
  odoo_connected: boolean;
  read_only: boolean;
  company_context: OdooCompanyContext;
  user_mapping: OdooUserMapping | null;
}

export interface OdooCompanyContext {
  selected: boolean;
  odoo_company_id?: number | null;
  odoo_company_name?: string | null;
  is_default?: boolean;
  selected_at?: string | null;
  message?: string | null;
}

export interface OdooSummary {
  customers: number;
  products: number;
  open_invoices: number;
  overdue_invoices: number;
  quotations: number;
  opportunities: number;
  projects: number;
  connected: boolean;
  company_id?: number | null;
  company_name?: string | null;
}

export interface OdooCustomer {
  id: number;
  name: string;
  email?: string | null;
  phone?: string | null;
  vat?: string | null;
  city?: string | null;
  is_company: boolean;
}

export interface OdooProduct {
  id: number;
  name: string;
  default_code?: string | null;
  list_price: string;
  standard_price: string;
  qty_available: number;
  uom?: string | null;
}

export interface OdooVendor {
  id: number;
  name: string;
  email?: string | null;
  phone?: string | null;
  city?: string | null;
  vat?: string | null;
}

export interface OdooSaleHistoryItem {
  id: number;
  order_name: string;
  order_date?: string | null;
  partner_id?: number | null;
  partner_name: string;
  product_id?: number | null;
  product_name: string;
  quantity: number;
  unit_price: string;
  subtotal: string;
  margin?: string | null;
  margin_pct?: number | null;
}

export interface OdooInvoice {
  id: number;
  name: string;
  partner_name: string;
  invoice_date?: string | null;
  due_date?: string | null;
  amount_total: string;
  amount_residual: string;
  currency: string;
  state: string;
  payment_state?: string | null;
}

export interface OdooQuotation {
  id: number;
  name: string;
  partner_id?: number | null;
  partner_name: string;
  date_order?: string | null;
  amount_total: string;
  currency?: string;
  state: string;
  user_id?: number | null;
  salesperson_name?: string | null;
  company_id?: number | null;
  company_name?: string | null;
  validity_date?: string | null;
}

export interface OdooQuotationLine {
  product_id?: number | null;
  product_name: string;
  description?: string | null;
  quantity: number;
  price_unit: string;
  discount: number;
  subtotal: string;
  taxes: string[];
}

export interface OdooQuotationDetail extends OdooQuotation {
  lines: OdooQuotationLine[];
  connected: boolean;
  message?: string | null;
}

export interface OdooOpportunity {
  id: number;
  name: string;
  partner_name?: string | null;
  expected_revenue: string;
  probability: number;
  stage?: string | null;
  date_deadline?: string | null;
}

export interface OdooProject {
  id: number;
  name: string;
  partner_name?: string | null;
  stage?: string | null;
}

export interface OdooListResponse<T = unknown> {
  items: T[];
  total: number;
  connected: boolean;
  message?: string | null;
}

export interface OdooQueryResponse {
  question: string;
  answer: string;
  data: Record<string, unknown>;
  query_type: string;
}

export interface OdooInvoiceLine {
  id: number;
  product_id?: number | null;
  product_name: string;
  quantity: number;
  unit_price: string;
  discount: number;
  tax_names?: string | null;
  subtotal: string;
  cost?: string | null;
  margin?: string | null;
  margin_pct?: number | null;
}

export interface OdooCustomerDetail {
  id: number;
  name: string;
  email?: string | null;
  phone?: string | null;
  vat?: string | null;
  city?: string | null;
  salesperson?: string | null;
  company_name?: string | null;
  connected: boolean;
  open_invoices: OdooInvoice[];
  overdue_invoices: OdooInvoice[];
  quotations: OdooQuotation[];
  opportunities: OdooOpportunity[];
  projects: OdooProject[];
  sales_history: OdooSaleHistoryItem[];
  products_purchased: { product_id: number; product_name: string; last_price?: string | null; average_price?: string | null; total_qty: number; sales_count: number }[];
  total_sales_historical: string;
  total_due: string;
  total_overdue: string;
  message?: string | null;
}

export interface OdooProductDetail {
  id: number;
  name: string;
  default_code?: string | null;
  category?: string | null;
  list_price: string;
  standard_price: string;
  qty_available: number;
  uom?: string | null;
  connected: boolean;
  sales_history: OdooSaleHistoryItem[];
  purchase_history: { id: number; vendor_name: string; order_name: string; order_date?: string | null; quantity: number; unit_price: string; subtotal: string }[];
  buyers: { partner_id: number; partner_name: string; last_price?: string | null; average_price?: string | null; sales_count: number }[];
  last_price?: string | null;
  average_price?: string | null;
  min_price?: string | null;
  max_price?: string | null;
  estimated_margin_pct?: number | null;
  message?: string | null;
}

export interface OdooInvoiceDetail {
  id: number;
  name: string;
  partner_id: number;
  partner_name: string;
  invoice_date?: string | null;
  due_date?: string | null;
  amount_total: string;
  amount_residual: string;
  currency: string;
  state: string;
  payment_state?: string | null;
  lines: OdooInvoiceLine[];
  total_margin?: string | null;
  margin_pct?: number | null;
  odoo_url?: string | null;
  connected: boolean;
  message?: string | null;
}

export interface OdooVendorDetail {
  id: number;
  name: string;
  email?: string | null;
  phone?: string | null;
  vat?: string | null;
  city?: string | null;
  connected: boolean;
  purchase_history: { id: number; order_name: string; product_name: string; quantity: number; unit_price: string; subtotal: string }[];
  products_supplied: { product_id: number; product_name: string; last_cost?: string | null; average_cost?: string | null; total_qty: number; purchase_count: number }[];
  total_purchase_value: string;
  purchase_orders: { name: string; date: string; amount_total: string; state: string }[];
  message?: string | null;
}

export function formatOdooAmount(value: string | number | null | undefined): string {
  if (value === null || value === undefined || value === "") return "—";
  const num = typeof value === "string" ? parseFloat(value) : value;
  if (Number.isNaN(num)) return String(value);
  return new Intl.NumberFormat("es-DO", {
    style: "currency",
    currency: "DOP",
    minimumFractionDigits: 2,
  }).format(num);
}

File: backend/app/schemas/odoo.py
from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class OdooHealthResponse(BaseModel):
    connected: bool
    read_only: bool = True
    version: str | None = None
    database: str | None = None
    message: str
    active_company_id: int | None = None
    active_company_name: str | None = None


class OdooCompanyResponse(BaseModel):
    id: int
    name: str
    currency_id: int | None = None
    currency_name: str | None = None
    partner_id: int | None = None
    is_active: bool = True
    selected_by_current_user: bool = False


class OdooCompaniesResponse(BaseModel):
    items: list[OdooCompanyResponse]
    connected: bool = True
    message: str | None = None


class OdooCompanyContextResponse(BaseModel):
    selected: bool = False
    odoo_company_id: int | None = None
    odoo_company_name: str | None = None
    is_default: bool = True
    selected_at: datetime | None = None
    message: str | None = None


class OdooCompanyContextUpdate(BaseModel):
    odoo_company_id: int


class OdooUserMappingResponse(BaseModel):
    id: UUID | None = None
    tenant_id: UUID
    jaios_user_id: UUID
    odoo_user_id: int
    odoo_login: str
    odoo_partner_id: int | None = None
    allowed_company_ids: list[int] = Field(default_factory=list)
    default_company_id: int | None = None
    is_active: bool = True
    last_verified_at: datetime | None = None


class OdooLinkUserRequest(BaseModel):
    odoo_login: str


class OdooMeResponse(BaseModel):
    jaios_user_id: UUID
    jaios_email: str
    jaios_name: str
    odoo_connected: bool
    read_only: bool = True
    company_context: OdooCompanyContextResponse
    user_mapping: OdooUserMappingResponse | None = None


class OdooSummaryResponse(BaseModel):
    customers: int = 0
    products: int = 0
    open_invoices: int = 0
    overdue_invoices: int = 0
    quotations: int = 0
    opportunities: int = 0
    projects: int = 0
    connected: bool = False
    company_id: int | None = None
    company_name: str | None = None


class OdooCustomerResponse(BaseModel):
    id: int
    name: str
    email: str | None = None
    phone: str | None = None
    vat: str | None = None
    city: str | None = None
    is_company: bool = False


class OdooProductResponse(BaseModel):
    id: int
    name: str
    default_code: str | None = None
    list_price: Decimal = Decimal("0")
    standard_price: Decimal = Decimal("0")
    qty_available: float = 0
    uom: str | None = None


class OdooSaleHistoryItem(BaseModel):
    id: int
    order_name: str
    order_date: str | None = None
    partner_id: int | None = None
    partner_name: str
    product_id: int | None = None
    product_name: str
    quantity: float
    unit_price: Decimal
    subtotal: Decimal
    margin: Decimal | None = None
    margin_pct: float | None = None


class OdooLastPriceResponse(BaseModel):
    product_name: str | None = None
    partner_name: str | None = None
    unit_price: Decimal | None = None
    min_price: Decimal | None = None
    max_price: Decimal | None = None
    order_date: str | None = None
    order_name: str | None = None
    average_price: Decimal | None = None
    sales_count: int = 0


class OdooVendorResponse(BaseModel):
    id: int
    name: str
    email: str | None = None
    phone: str | None = None
    city: str | None = None
    vat: str | None = None


class OdooProductPurchaseItem(BaseModel):
    id: int
    vendor_name: str
    order_name: str
    order_date: str | None = None
    quantity: float
    unit_price: Decimal
    subtotal: Decimal


class OdooProductSuppliedItem(BaseModel):
    product_id: int
    product_name: str
    last_cost: Decimal | None = None
    average_cost: Decimal | None = None
    total_qty: float = 0
    purchase_count: int = 0


class OdooInvoiceResponse(BaseModel):
    id: int
    name: str
    partner_name: str
    invoice_date: str | None = None
    due_date: str | None = None
    amount_total: Decimal
    amount_residual: Decimal
    currency: str = "DOP"
    state: str
    payment_state: str | None = None


class OdooQuotationLineResponse(BaseModel):
    product_id: int | None = None
    product_name: str
    description: str | None = None
    quantity: float = 0
    price_unit: Decimal = Decimal("0")
    discount: float = 0
    subtotal: Decimal = Decimal("0")
    taxes: list[str] = Field(default_factory=list)


class OdooQuotationResponse(BaseModel):
    id: int
    name: str
    partner_id: int | None = None
    partner_name: str
    date_order: str | None = None
    amount_total: Decimal
    currency: str = "DOP"
    state: str
    user_id: int | None = None
    salesperson_name: str | None = None
    company_id: int | None = None
    company_name: str | None = None
    validity_date: str | None = None


class OdooQuotationDetailResponse(OdooQuotationResponse):
    lines: list[OdooQuotationLineResponse] = Field(default_factory=list)
    connected: bool = True
    message: str | None = None


class OdooQuotationSearchParams(BaseModel):
    q: str | None = None
    quotation_number: str | None = None
    customer: str | None = None
    salesperson: str | None = None
    product: str | None = None
    date_from: str | None = None
    date_to: str | None = None
    amount_min: Decimal | None = None
    amount_max: Decimal | None = None
    state: str | None = None
    company_id: int | None = None
    limit: int = 50


class OdooOpportunityResponse(BaseModel):
    id: int
    name: str
    partner_name: str | None = None
    expected_revenue: Decimal = Decimal("0")
    probability: float = 0
    stage: str | None = None
    date_deadline: str | None = None


class OdooProjectResponse(BaseModel):
    id: int
    name: str
    partner_name: str | None = None
    stage: str | None = None


class OdooTicketResponse(BaseModel):
    id: int
    name: str
    partner_name: str | None = None
    stage: str | None = None
    priority: str | None = None


class OdooPurchaseHistoryItem(BaseModel):
    id: int
    order_name: str
    order_date: str | None = None
    partner_name: str
    product_name: str
    quantity: float
    unit_price: Decimal
    subtotal: Decimal


class OdooQueryResponse(BaseModel):
    question: str
    answer: str
    data: dict[str, Any] = Field(default_factory=dict)
    query_type: str


class OdooListResponse(BaseModel):
    items: list[Any]
    total: int
    connected: bool = True
    message: str | None = None
    company_id: int | None = None


class OdooInvoiceLineResponse(BaseModel):
    id: int
    product_id: int | None = None
    product_name: str
    quantity: float
    unit_price: Decimal
    discount: float = 0
    tax_names: str | None = None
    subtotal: Decimal
    cost: Decimal | None = None
    margin: Decimal | None = None
    margin_pct: float | None = None


class OdooProductPurchasedItem(BaseModel):
    product_id: int
    product_name: str
    last_price: Decimal | None = None
    average_price: Decimal | None = None
    total_qty: float = 0
    sales_count: int = 0


class OdooCustomerDetailResponse(BaseModel):
    id: int
    name: str
    email: str | None = None
    phone: str | None = None
    vat: str | None = None
    city: str | None = None
    salesperson: str | None = None
    company_name: str | None = None
    connected: bool = True
    open_invoices: list[OdooInvoiceResponse] = Field(default_factory=list)
    overdue_invoices: list[OdooInvoiceResponse] = Field(default_factory=list)
    quotations: list[OdooQuotationResponse] = Field(default_factory=list)
    opportunities: list[OdooOpportunityResponse] = Field(default_factory=list)
    projects: list[OdooProjectResponse] = Field(default_factory=list)
    sales_history: list[OdooSaleHistoryItem] = Field(default_factory=list)
    products_purchased: list[OdooProductPurchasedItem] = Field(default_factory=list)
    total_sales_historical: Decimal = Decimal("0")
    total_due: Decimal = Decimal("0")
    total_overdue: Decimal = Decimal("0")
    message: str | None = None


class OdooProductBuyerItem(BaseModel):
    partner_id: int
    partner_name: str
    last_price: Decimal | None = None
    average_price: Decimal | None = None
    sales_count: int = 0


class OdooProductDetailResponse(BaseModel):
    id: int
    name: str
    default_code: str | None = None
    category: str | None = None
    list_price: Decimal = Decimal("0")
    standard_price: Decimal = Decimal("0")
    qty_available: float = 0
    uom: str | None = None
    connected: bool = True
    sales_history: list[OdooSaleHistoryItem] = Field(default_factory=list)
    purchase_history: list[OdooProductPurchaseItem] = Field(default_factory=list)
    buyers: list[OdooProductBuyerItem] = Field(default_factory=list)
    last_price: Decimal | None = None
    average_price: Decimal | None = None
    min_price: Decimal | None = None
    max_price: Decimal | None = None
    estimated_margin_pct: float | None = None
    related_invoices: list[OdooInvoiceResponse] = Field(default_factory=list)
    related_quotations: list[OdooQuotationResponse] = Field(default_factory=list)
    message: str | None = None


class OdooInvoiceDetailResponse(BaseModel):
    id: int
    name: str
    partner_id: int
    partner_name: str
    invoice_date: str | None = None
    due_date: str | None = None
    amount_total: Decimal
    amount_residual: Decimal
    currency: str = "DOP"
    state: str
    payment_state: str | None = None
    lines: list[OdooInvoiceLineResponse] = Field(default_factory=list)
    total_margin: Decimal | None = None
    margin_pct: float | None = None
    odoo_url: str | None = None
    connected: bool = True
    message: str | None = None


class OdooVendorDetailResponse(BaseModel):
    id: int
    name: str
    email: str | None = None
    phone: str | None = None
    vat: str | None = None
    city: str | None = None
    connected: bool = True
    purchase_history: list[OdooPurchaseHistoryItem] = Field(default_factory=list)
    products_supplied: list[OdooProductSuppliedItem] = Field(default_factory=list)
    total_purchase_value: Decimal = Decimal("0")
    purchase_orders: list[dict[str, Any]] = Field(default_factory=list)
    message: str | None = None
